src/main/scala/firrtl/transforms/FixAddingNegativeLiteralsTransform.scala

```diff
@@ -0,0 +1,103 @@
+// See LICENSE for license details.
+
+package firrtl.transforms
+
+import firrtl.{CircuitState, LowForm, Namespace, PrimOps, Transform, Utils, WRef}
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.PrimOps.{Add, AsSInt, Sub, Tail}
+
+import scala.collection.mutable
+
+object FixAddingNegativeLiterals {
+
+  /** Returns the maximum negative number represented by given width
+    * @param width width of the negative number
+    * @return maximum negative number
+    */
+  def minNegValue(width: BigInt): BigInt = -(1 << (width.toInt - 1))
+
+  /** Updates the type of the DoPrim from its arguments (e.g. if is UnknownType)
+    * @param d input DoPrim
+    * @return updated DoPrim with calculated type
+    */
+  def setType(d: DoPrim): DoPrim = {
+    PrimOps.set_primop_type(d)
+  }
+
+  /** Returns a module with fixed additions of negative literals
+    * @param m input module
+    * @return updated module
+    */
+  def fixupModule(m: DefModule): DefModule = {
+    val namespace = Namespace(m)
+    m map fixupStatement(namespace)
+  }
+
+  /** Returns a statement with fixed additions of negative literals
+    * @param namespace object to enabling creating unique names
+    * @param s input statement
+    * @return updated statement
+    */
+  def fixupStatement(namespace: Namespace)(s: Statement): Statement = {
+    val stmtBuffer = mutable.ListBuffer[Statement]()
+    val ret = s map fixupStatement(namespace) map fixupExpression(Utils.get_info(s), namespace, stmtBuffer)
+    if(stmtBuffer.isEmpty) {
+      ret
+    } else {
+      stmtBuffer += ret
+      Block(stmtBuffer.toList)
+    }
+  }
+
+  /** Returns a statement with fixed additions of negative literals
+    * @param info Info of statement containing this expression
+    * @param namespace object to enabling creating unique names
+    * @param stmtBuffer mutable buffer of statements - append to this for it to be inlined in the module
+    * @param e expression to fixup
+    * @return fixed expression
+    */
+  def fixupExpression(info: Info, namespace: Namespace, stmtBuffer: mutable.ListBuffer[Statement])
+                     (e: Expression): Expression = {
```

Comment:
I thought I already left this comment but I can't find it so sorry if you've addressed this.

I don't like mutation being part of the public API. Why not make the public version return a tuple of `Seq[Statement]` and `Expression` and have the version with mutation be a private inner function?

Reply:
I can, but then the application of the function in `fixupStatement` will be less elegant. I'll add a public version which is immutable and make the existing function private and only called in `fixupStatement`.

Reply:
My issue is with mutation in the public API, you could make `fixupExpression` private, you could make it have a private form that has mutation, but I really don't think we want more mutation on public APIs

Reply:
Which I see is what you did 🙂 Thanks Adam!